rfcs/0000-changes-lifecyle.md

```diff
@@ -0,0 +1,318 @@
+- Feature Name: changes-lifecycle
+- Start Date: 2024-08-05
+- RFC PR: [FuelLabs/sway-rfcs#41](https://github.com/FuelLabs/sway-rfcs/pull/41)
+
+# Summary
+
+[summary]: #summary
+
+This RFC is a guide on how to introduce breaking changes following the compiler release lifecycle.
+
+# Motivation
+
+[motivation]: #motivation
+
+To maximize user experience when updating the compiler, we need a guide on how to introduce breaking changes.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The compiler will follow a "rolling release" scheme, which means that periodically (to be specified) a
+new major version will be released.
+
+This means that as soon as the compiler reaches version "1.0.0", the next **major version** will be "1.1.0"; and
```

Comment:
Hum...  I think `major.minor.patch` makes more sense. Not sure what would be a major increase in our case.

Reply:
I changed to a more neutral wording, just calling for "next version" and "next patch".

Reply:
This section is possibly the most important part of the document as it deals with how we actually handle the versioning of breaking changes, but I think I'm still a little confused. I'm hoping the following example can help me understand:

Let's say we're currently live at version `X.Y.Z`, and we have a new release ready to go containing the following in its release notes:

```
### Breaking Changes

- This release removes the ABC feature
```

Seeing as the next release will be a breaking change, what will be the SemVer for its release? Will it be:

- (X+1).Y.Z
- X.(Y+1).Z
- X.Y.(Z+1)

In SemVer this release will be `(X+1).Y.Z` but as I read this current section the rolling release will put it at `X.(Y+1).Z`? Whatever the version we use is not my point here - I'm just thinking we need to be more explicit here with whichever way we decide so that it's not ambiguous in the future.

To be more explicit, maybe a new sub-section should be added here called "Semantic Versioning" which summarises how we will categorise/number each type of change i.e something like:

### Semantic Versioning

The compiler will follow a "rolling release" scheme, meaning that each new release (happening periodically but yet to be specified) will steamroll over all previous releases.

#### MAJOR

`MAJOR` number changes are strictly reserved to deal with dramatic differences in the Sway Language.

#### MINOR

Every normal change, which includes **_all_** breaking changes, new features, and bug fixes, will be a bump increase in the `MINOR` number.

#### PATCH

`PATCH` number changes are out-of-band, and are strictly reserved for security fixes to each currently supported `MAJOR.MINOR` release. In other words, if a security fix is needed, every existing supported release will include a new patch bump release.
```

Reply:
I like @alfiedotwtf's detailed wording on major, minor, and patch. Only the title "Semantic Versioning" I find potentially confusing because, at least in my head, it immediately associates on semver where the expectation is to increase the major on breaking changes while in our case, the major version is indeed more a "marketing version". Perhaps just title it "Versioning Scheme"?